unsafe-fields/src/lib.rs

```diff
@@ -215,33 +219,53 @@ impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
     /// The caller is responsible for upholding any safety invariants associated
     /// with this field.
     #[inline(always)]
-    pub const unsafe fn as_ref(&self) -> &F {
+    pub const unsafe fn as_ref_unchecked(&self) -> &F {
         // SAFETY: This method is unsafe to call.
         &self.field
     }
 
+    /// Gets a reference to the inner value safely so long as the inner value is
+    /// immutable.
+    ///
+    /// If [`F: Immutable`][immutable], then `F` does not permit interior
+    /// mutation, and so it is safe to return a reference to it.
+    ///
+    /// [immutable]: zerocopy_0_8::Immutable
+    #[inline(always)]
+    #[cfg(feature = "zerocopy_0_8")]
+    #[cfg_attr(doc_cfg, doc(cfg(feature = "zerocopy_0_8")))]
+    pub const fn as_ref(&self) -> &F
+    where
+        F: zerocopy_0_8::Immutable,
+    {
+        // SAFETY: `F: Immutable` guarantees that the returned `&F` cannot be
+        // used to mutate `self`, and so it cannot be used to violate any
+        // invariant.
+        unsafe { self.as_ref_unchecked() }
+    }
+
     /// Gets a mutable reference to the inner value.
     ///
     /// # Safety
     ///
     /// The caller is responsible for upholding any safety invariants associated
     /// with this field.
     #[inline(always)]
-    pub unsafe fn as_mut(&mut self) -> &mut F {
+    pub unsafe fn as_mut_unchecked(&mut self) -> &mut F {
```

Comment:
Same concern as `new`. I don't *think* there's a safe version we could possibly offer here that we'd want to distinguish the unsafe version from.

Reply:
Done.